libtracepoint-control-cpp/src/TracingSession.cpp

```diff
@@ -0,0 +1,896 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <tracepoint/TracingSession.h>
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+
+#include <unistd.h>
+#include <poll.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <asm/unistd.h> // __NR_perf_event_open
+#include <linux/perf_event.h>
+
+#ifdef NDEBUG
+#define DEBUG_PRINTF(...) ((void)0)
+#else // NDEBUG
+#include <stdio.h>
+#define DEBUG_PRINTF(...) printf(__VA_ARGS__)
+#endif // NDEBUG
+
+using namespace std::string_view_literals;
+using namespace tracepoint_control;
+using namespace tracepoint_decode;
+
+static long
+perf_event_open(
+    struct perf_event_attr* pe,
+    pid_t pid,
+    int cpuIndex,
+    int groupFd,
+    unsigned long flags) noexcept
+{
+    return syscall(__NR_perf_event_open, pe, pid, cpuIndex, groupFd, flags);
+}
+
+// Return the smallest power of 2 that is >= pageSize and >= bufferSize.
+// Assumes pageSize is a power of 2.
+static size_t
+RoundUpBufferSize(uint32_t pageSize, size_t bufferSize) noexcept
```

Comment:
Could be constexper right?

Reply:
I guess. But none of the callers will be passing constant expressions into it (pageSize, in particular, is obtained from an API call), so I don't think it would do any good.